src/acrn/acrn_driver.c

```diff
@@ -2556,6 +2556,50 @@ acrnInitPlatform(acrnPlatformInfoPtr pi, virNodeInfoPtr nodeInfo,
     return ret;
 }
 
+static int
+acrnPersistentDomainInit(virDomainObjPtr dom, void *opaque)
+{
+    int ret = -1;
+    acrnConnectPtr privconn = opaque;
+
+    unsigned char hvUUID[VIR_UUID_BUFLEN];
+    struct acrnVmList *vmList;
+    acrnDomainObjPrivatePtr priv;
+    virObjectEventPtr event = NULL;
+
+    vmList = acrnVmListNew();
+    if (!vmList)
+        goto cleanup;
+
+    acrnDriverLock(privconn);
```

Comment:
Please remove this. Since the domain's rw lock is held when this function is called, calling acrnDriverLock() creates a potential deadlock.

Reply:
Removing the lock/unlock in `acrnPersistentDomainInit` should not result in any problems, since here we are in the initialization phase of the driver. The problem obviously is holding a domain lock and then acquiring the driver lock. What about removing lock/unlock in `acrnPersistentDomainInit` and acquire the driver lock outside the iteration:
```
acrnDriverLock(privconn);
virDomainObjListForEach(acrn_driver->domains, ...)
acrnDriverUnlock(privconn);
```

It took my me some additional steps through the code to understand your point about the recursive locking (using `virDomainObjListForEach` in `acrnPersistentDomainInit` and recursively in `acrnAllocateVm`). Are there any chances for a clean way out of this? Probably adding a 'lockless' variant of `acrnAllocateVm` using `virHashForEach` directly protected just by the driver lock?  Or just doing this in `acrnAllocateVm` as it is? Since right now the driver lock is held by every caller of `acrnAllocateVm` using `virHashForEach` is possible, it means just adding the requirement to hold the driver lock when calling `acrnAllocateVm` for any future user. What do you think about it?

Reply:
Just minutes later, the above seems a bad idea since this requires the driver lock to be held everywhere the domain list gets manipulated!

Reply:
Just found this in the manpage of pthread_rwlock_rdlock:

> A thread may hold multiple concurrent read locks on rwlock (that is, successfully call the pthread_rwlock_rdlock() function n times). If so, the application shall ensure that the thread performs matching unlocks (that is, it calls the pthread_rwlock_unlock() function n times).

Reply:
We could try to have a lockless flavor of `acrnAllocateVm`. Right now, it would only be used in `acrnStateInitialize` so we don't need to grab the driver lock at all (`acrn_driver` is not yet observable to anyone other than this function). In this case, I'd actually prefer your first version using `notify` in `virDomainObjListLoadAllConfigs`. The domain's write lock will be held during `notify`, so it can't be acquired (read or write) again in `acrnAllocateVm`.